actors/multisig/tests/multisig_actor_test.rs

```diff
@@ -1090,46 +1090,481 @@ fn test_signer_swap() {
+        let h = util::ActorHarness::new();
+        rt.set_balance(send_value.clone());
+        rt.set_received(TokenAmount::zero());
+        h.construct_and_verify(&mut rt, 2, 0, 0, signers);
+
+        let fake_params = RawBytes::from(vec![1, 2, 3, 4]);
+        let fake_method = 42;
+        rt.set_caller(*ACCOUNT_ACTOR_CODE_ID, anne);
+        h.propose_ok(&mut rt, chuck, send_value.clone(), fake_method, fake_params.clone());
+        let bad_hash = compute_proposal_hash(
+            &Transaction {
+                to: chuck,
+                value: send_value.clone(),
+                method: fake_method,
+                params: fake_params.clone(),
+                approved: vec![bob], //mismatch
+            },
+            &rt,
+        )
+        .unwrap();
+        rt.set_caller(*ACCOUNT_ACTOR_CODE_ID, bob);
+        expect_abort(ExitCode::USR_ILLEGAL_ARGUMENT, h.approve(&mut rt, TxnID(0), bad_hash));
+    }
+
+    #[test]
+    fn accept_approval_with_no_proposal_hash() {
+        let msig = Address::new_id(100);
+        let anne = Address::new_id(101);
+        let bob = Address::new_id(102);
+        let chuck = Address::new_id(103);
+        let signers = vec![anne, bob];
+        let mut rt = construct_runtime(msig);
+        let send_value = TokenAmount::from(10u8);
+        let h = util::ActorHarness::new();
+        rt.set_balance(send_value.clone());
+        rt.set_received(TokenAmount::zero());
+        h.construct_and_verify(&mut rt, 2, 0, 0, signers);
+
+        let fake_params = RawBytes::from(vec![1, 2, 3, 4]);
+        let fake_method = 42;
+        rt.set_caller(*ACCOUNT_ACTOR_CODE_ID, anne);
         h.propose_ok(&mut rt, chuck, send_value.clone(), fake_method, fake_params.clone());
 
-    // assert txn
-    let expect_txn = Transaction {
-        to: chuck,
-        value: send_value.clone(),
-        method: fake_method,
-        params: fake_params.clone(),
-        approved: vec![anne],
-    };
-    h.assert_transactions(&rt, vec![(TxnID(0), expect_txn)]);
+        rt.set_caller(*ACCOUNT_ACTOR_CODE_ID, bob);
+        rt.expect_send(
+            chuck,
+            fake_method,
+            fake_params,
+            send_value,
+            RawBytes::default(),
+            ExitCode::OK,
+        );
+        rt.expect_validate_caller_type(vec![*ACCOUNT_ACTOR_CODE_ID, *MULTISIG_ACTOR_CODE_ID]);
+        let params = TxnIDParams { id: TxnID(0), proposal_hash: Vec::<u8>::new() };
+        rt.call::<MultisigActor>(Method::Approve as u64, &RawBytes::serialize(params).unwrap())
+            .unwrap();
+        rt.verify();
+    }
 
-    // approval
-    rt.set_balance(send_value.clone());
-    rt.set_caller(*ACCOUNT_ACTOR_CODE_ID, bob);
-    rt.expect_send(chuck, fake_method, fake_params, send_value, fake_ret, ExitCode::OK);
-    h.approve_ok(&mut rt, TxnID(0), proposal_hash);
-    h.assert_transactions(&rt, vec![]);
+    #[test]
+    fn fail_approve_tx_more_than_once() {
```

Comment:
I think it's worth splitting this into:

- X proposes a tx, X fails to approve it (what this test is currently doing)
- X proposes a tx, Y approves it, Y fails to approve it again (would need an msig with a threshold > 2)

Reply:
This falls under the category of new test.  This is out of scope for our current work which is aiming to achieve parity with specs-actors testing across all actors before expanding unit tests.  However I like the idea of recording these somewhere so I opened this issue to do that: https://github.com/filecoin-project/builtin-actors/issues/314